### PR TITLE
joe.covariances.port

### PR DESCRIPTION
**Pull request by jdramsey**
Porting cleaned up covariance classes.

### CODE REVIEW

**Comment by jdramsey** [comment]
Here's what I did. I had covariance classes in another branch, including Kevin's implementations. I cleaned these up, put them behind a common interface. I revised the CovarianceMatrix class to calculate covariances using whatever method the user selects, by default Kevin's method for doubles with bias correction. This can be changed. More calculation methods can be added easily.

I changed all uses of CovarianceMatrixOnTheFly to CovarianceMatrix and removed that class.

This incurred a number of unit test failures, which I fixed; they were all for dumb bugs.

Fix for https://github.com/cmu-phil/tetrad/issues/1175.

**Comment by kvb2univpitt** [comment]
Ok.  I am checking it now.

File: tetrad-lib/src/main/java/edu/cmu/tetrad/data/CovariancesFloatStandard.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+        for (int i = 0; i < _data.length; i++) {
+            for (int j = 0; j < _data[0].length; j++) {
+                this._data[i][j] = (float) (round(data[i][j] * 100.0) / 100.0);
+            }
+        }
+
+        this.numOfRows = data.length;
+        this.numOfCols = data[0].length;
+        this.covariances = compute(biasCorrected);
```
Comment (kvb2univpitt) [inline tetrad-lib/src/main/java/edu/cmu/tetrad/data/CovariancesFloatStandard.java]: The compute method should be private if called inside of constructor.  If not, the user can call it again.

File: tetrad-lib/src/main/java/edu/cmu/tetrad/data/CovariancesFloatStandard.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    @Override
+    public double covariance(int i, int j) {
+        return covariances[i][j];
+    }
+
+    public int size() {
+        return numOfCols;
+    }
+
+    public void setCovariance(int i, int j, double v) {
```
Comment (kvb2univpitt) [inline tetrad-lib/src/main/java/edu/cmu/tetrad/data/CovariancesFloatStandard.java]: This method should not exist.  A user should not be able to modify the covariance matrix.  This will lead to data corruption.

**Comment by kvb2univpitt** [comment]
@jdramsey I'll work on those issues that I mentioned above from your branch.

**Comment by jdramsey** [comment]
Oh ok. Sorry I was going to get to them. I got busy with another project

**Comment by kvb2univpitt** [comment]
@jdramsey I am going to do  some refactoring.  Basically, I am going to remove CovariancesFloatStandard and just replace with edu.cmu.tetrad.stat.correlation.CovarianceMatrix.  No need to have the same implementation in two different places.  Also, alternative covariance is basically compute the covariance on double data and return float data.  Do we even need an "alternative" covariance?

**Comment by jdramsey** [comment]
I think we need it--alternative covariance is your method.

**Comment by jdramsey** [comment]
Oh, why do we need two? Because I don't know which performs the best yet.

**Comment by kvb2univpitt** [comment]
@jdramsey 

> Oh, why do we need two? Because I don't know which performs the best yet.

Ok.  Now I see the difference.  Well, in that case, I am just going to leave everything the way it is.  I thought it was the same implementation.

**Comment by kvb2univpitt** [comment]
@espinoj I am going to leave everything the way it is.  Do you want to include this change for this coming release?

**Comment by jdramsey** [comment]
Yes, though if you want to add a parallelized version of your algorithm, that would be fine.

Basically, everything in the code should use the same covariance method. That way if we find issues, we can fix them all in one place. That was my idea.

It's OK to leave the choice of covariance method hard-coded for the moment. My issue is, does one of the other lead to better algorithm performance? I still haven't sorted that out, though I spent a fair amount of time on it. I think the double method perform better than the float method. But once this is in development I will do a study comparing the methods all to one another and we will know. Then if we agree, we can remove the methods that are inferior.

Basically, I will use the algorithm comparison tool to run searches for PC and FGES with the different options head to head and see which one does the best. I think it's just heuristic.

**Comment by jdramsey** [comment]
But I'd rather do it while it's in the development branch, just so I don't introduce a bug from my current poison branch.

**Comment by kvb2univpitt** [comment]
@jdramsey I'll add the parallelized version of my implementation to this branch.

**Comment by kvb2univpitt** [comment]
@jdramsey I am going to hold off on merging this into the development branch.  Jeremy is going to make a release this week.  Since we haven't fully tested the new covariance matrices, we won't include it.

**Comment by jdramsey** [comment]
Sounds good! I'll test it in that branch then. @kvb2univpitt

**Comment by jdramsey** [comment]
Canceling this, at Kevin's request.
